// @telepilotco/tdl/dist/addon.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.loadAddon = loadAddon;
var _path = _interopRequireDefault(require("path"));
var _nodeGypBuild = _interopRequireDefault(require("node-gyp-build"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const packageDir = _path.default.join(__dirname, '..');
let loaded = false;
function loadAddon(libraryFile) {
  if (loaded) throw Error('The node addon is already loaded');
  const addon = (0, _nodeGypBuild.default)(packageDir);
  addon.loadTdjson(libraryFile);
  loaded = true;
  return {
    create: addon.create,
    send: addon.send,
    receive: addon.receive,
    execute: addon.execute,
    destroy: addon.destroy,
    setLogFatalErrorCallback: addon.setLogFatalErrorCallback,
    setLogMessageCallback: addon.setLogMessageCallback,
    tdn: {
      init: addon.tdnInit,
      createClientId: addon.tdnCreateClientId,
      send: addon.tdnSend,
      receive: addon.tdnReceive,
      execute: addon.tdnExecute
    }
  };
}
